### DISCUSSION TITLE
Proposal: remove propertyNames

### DISCUSSION THREAD

**Discussion by SorinGFS** [json-schema-org/community#768] [Specification]
The keyword `propertyNames` doesn't fit in `json-schema` algoritm and IMO should be removed.

### Motivation

- This keyword is one of 11 keywords that allow any schemas
- at the same time this keyword is applicable only to the names of properties of an object
- the data that is targeted by this keyword cannot be located in json, notations such as dataLocation cannot be produced
- the previously presented conditions force implementations to check the data type before applying a schema
- the previous problem basically forces implementations to validate almost any schema that is validated by a metaschema, thus annihilating the very existence of this keyword
- by introducing this artificial check for data type before applying a schema on those 11 keywords implementations are basically forced to validate potentially invalid schemas on those keywords too

I searched for the presence of this keyword in 2300 schemes, I found 109 instances of which about half were stupid like `type: string`, about a quarter were enum or not enum, and the others were pattern. None of those instances were something that could not be replaced by other schemas.

**Comment by jdesrosiers** [comment]
> * the data that is targeted by this keyword cannot be located in json, notations such as dataLocation cannot be produced

I think you mean that you can't reference it with a JSON Pointer. I've noticed this as well and have considered whether that means `propertyNames` belongs in JSON Schema. I'm not 100% sure, but I am convinced that it provides users with the ability to do things they couldn't without it and to express constraints in a cleaner way that they could without it.

> * the previously presented conditions force implementations to check the data type before applying a schema
> * the previous problem basically forces implementations to validate almost any schema that is validated by a metaschema, thus annihilating the very existence of this keyword
> * by introducing this artificial check for data type before applying a schema on those 11 keywords implementations are basically forced to validate potentially invalid schemas on those keywords too

None of this makes sense to me. I don't understand the problem you're trying to describe. Whatever it is, I don't think I've encountered it. In my implementation, the only issue I encountered was that JSON Pointer can't point to a property name, but that's only a problem for the output format. It doesn't affect the algorithm at all.

**Reply by gregsdennis** [reply]
> JSON Pointer can't point to a property name

(Just for clarity...)

While technically true, it can still indicate a property name given the context that we're looking at property names in the first place.

For example, if I have

```json
{
  "propertyNames": {
    "pattern": "^foo[0-9]+$"
  }
}
```

with the instance `{"bar": true}`, I'd get an output object that includes

```jsonc
{
  "evaluationPath": "/propertyNames",
  "instanceLocation": "/bar"
  // ...
}
```

With these two things, I have the context that I'm looking at property names, so I can infer that `bar` is a property name that doesn't match the subschema at `/propertyNames`.  It's obvious that it's not looking at the `true` value.

I see no problem with this.

**Reply by SorinGFS** [reply]
>  It's obvious that it's not looking at the true value. I see no problem with this.

I think you created the annotation system for a reason. And I think you expected that annotations to be able to provide programmatic access to those references for things like code generators. How would be this goal accomplished having exceptions?

**Reply by SorinGFS** [reply]
... continuation

basically the same reference for dataLocation would be shared by 2 purposes, in order to be used by a program would require some extra tests to "detect" what you meant by that reference. All these problems (a check for data instance of object in validators, and the checks to "detect" for what is that data reference) are created by this `propertyNames` keyword which is the only keyword that doesn't work like all the others.

**Comment by SorinGFS** [comment]
> but I am convinced that it provides users with the ability to do things they couldn't without it and to express constraints in a cleaner way that they could without it.

In 2300 schema files from schemastore I couldn't find anything like that. Of course, it also depends on what you mean by clean, of course propertyNames is the cleanest, only that it allows even simple errors, and from my point of view the following are quite clean:


**for allowing exact names or patterns (with desired pattern instead of this in example):**
```json
{
  "$id": "https://json-schema.hyperjump.io/schema",
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "type":"object",
  "patternProperties": {"^(foo|bar)$": true},
  "additionalProperties": false
}
```
**Note:** allowing property names without defining a schema for the values is another non-sense IMO, the proof of that is the rare use of this `propertyNames` keyword.

**for deniying names:**
```json
{
  "$id": "https://json-schema.hyperjump.io/schema",
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "type":"object",
  "not": {"anyOf": [ {"required": ["foo"]}, {"required": ["bar"]}]}
}
```

> Whatever it is, I don't think I've encountered it. In my implementation

Now take a look at this schema, I used your implementation, but I'm sure would produce the same result with any implementation due to the reasons I presented regarding the algorithm:

```json
{
  "$id": "https://json-schema.hyperjump.io/schema",
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "type":"object",
  "propertyNames": {"propertyNames":{"type":"boolean"}},
  "allOf": [{"propertyNames": {"propertyNames":{"type":"boolean"}}}]
}
```

A non-sense schema like this is perfectly valid and would validate any object. And this is just a quick example, basically any schema valid against metaschema would validate the object. In fact this is a broader problem in `json-schema` (the lack of restriction for those 11 keywords) which is not visible because the metaschema is correct, but schema authors are not as knowledgeable as you are.

> the only issue I encountered was that JSON Pointer can't point to a property name, but that's only a problem for the output format. It doesn't affect the algorithm at all.

Without this "small" issue the algorithm would work without checking for `data instanceof Object` on those 11 keywords. To me this looks like workaround to fix this problem, not the solution. The problem commes from the "wildcard" power of this keyword which like `items` or `additionalProperties` or any of those 11 keywords are unrestricted about what they can contain. Using this `propertyNames` keyword with a wrong schema in it (and because the lack of restrictions in metaschema the probability of error is high) would validate the object, while using solutions like ones I showed at the top would always produce any desired outcome.

**Reply by notEthan** [reply]
`propertyNames` does require substantially different processing than any other subschema. As you point out, the instance it applies to isn't one that a pointer can reference in the document, rather the property name has to be considered as its own entity. But it is not unique in this: `contentSchema` is also a schema that describes an instance that is not directly in the document, but is the result of some parsing of the instance.

> * the data that is targeted by this keyword cannot be located in json, notations such as dataLocation cannot be produced

It is true that you need more than a pointer to get to the instance. You could point to the instance object and then additionally reference the property name, or as Greg suggests you can point to the property and infer that the property name is the last token since the schema is using the `propertyNames` keyword.

> * the previously presented conditions force implementations to check the data type before applying a schema

Many keywords require checking instance data type before they are used - what's different?

> I searched for the presence of this keyword in 2300 schemes, I found 109 instances of which about half were stupid like `type: string`, about a quarter were enum or not enum, and the others were pattern.

So about 5% of sampled schemas use it, half non-trivially - that's pretty significant, I would say, and breaking that many schemas is a hard one to sell.

> None of those instances were something that could not be replaced by other schemas.

You offer some examples with `patternProperties` and `required`/`not`/`anyOf` - these look to me much less readable and maintainable than equivalent schemas using `propertyNames`. I recognize that is a matter of opinion and you describe them as 'quite clean' ... I disagree.

> * the previous problem basically forces implementations to validate almost any schema that is validated by a metaschema, thus annihilating the very existence of this keyword

> * by introducing this artificial check for data type before applying a schema on those 11 keywords implementations are basically forced to validate potentially invalid schemas on those keywords too

> A non-sense schema like this is perfectly valid and would validate any object.

It has always been possible to write nonsense schemas that validate any instance or no instances, with many keywords. You can create constructions that are nonsense in all sorts of formats and languages - regular expressions, any programming language, English ... You have to rely on the author of a schema or any other work to make some sense. Taking away expressive power (e.g. with `propertyNames`) isn't the answer to that.

**Reply by SorinGFS** [reply]
> Many keywords require checking instance data type before they are used - what's different?

I finished my implementation without any data type check, except for this keyword. Not published yet. This is how I actually find the problem.

> So about 5% of sampled schemas use it, half non-trivially - that's pretty significant, I would say, and breaking that many schemas is a hard one to sell.

No. There are actually few schemas, I will count them and tell, I said 109 instances of keyword, but there are some schemas which contain lots of them.

**Reply by SorinGFS** [reply]
> But it is not unique in this: contentSchema is also a schema that describes an instance that is not directly in the document, but is the result of some parsing of the instance.

That is a different matter, I didn't tell anything about that. And `contentSchema`'s data location can be referenced: is ... ${dataLocation}/contentSchema

**Reply by SorinGFS** [reply]
@notEthan 

hard to count, has over a million rows, but if wou want to check yourself: (removed link)
You can search for the keyword and navigate one by one. Many of them are in asyncapi, a schema which has many design problems. Also, the `json-schema.org`'s keywords are also included in those 109

**Reply by notEthan** [reply]
> I finished my implementation without any data type check, except for this keyword. Not published yet. This is how I actually find the problem.

This is confusing to me. Many keywords state that they only apply to a particular data type, and many tests of the JSON Schema Test Suite check that these keywords have no effect on other data types.

> And `contentSchema`'s data location can be referenced: is ... ${dataLocation}/contentSchema

I may not have phrased the comparison as well as I hoped.
The instance described by `contentSchema` is different than the instance (a string) that the schema containing `contentSchema` applies to - you have to parse the content to get to that instance.
Comparably, the instance described by `propertyNames` is different than the instance (an object) that the schema containing `propertyNames` applies to - you have to identify a property name.
The point just being that `propertyNames` is not unique in needing more than a simple pointer to get to the instance the schema describes.

> There are actually few schemas, I will count them and tell

I see, I misunderstood the numbers. I don't know how low of a usage would qualify for removal. `propertyNames` seems worth having to me, on balance considering usage (pretty low), implementation complexity (moderate), and expressive power. I can certainly appreciate considering that balance to fall the other way. Considering also that it's been around since draft 6, though, I doubt that its downsides are enough for it to be removed.

**Reply by gregsdennis** [reply]
While low usage could be an argument against introducing a proposal into the spec, it absolutely does not warrant _removal_ from the spec.

Many of the keywords that we currently have are merely simpler, more concise ways of writing previously achieveable constraint sets.  An example of this is `if`/`then`/`else`, which was introduced as a more readable way of writing conditionals that could already be achieved with `anyOf`/`const` constructs.  So the argument that `propertyNames` can be achieved with other means also doesn't stand.

Providing a minimal keyword set that expresses all constraints isn't a goal of JSON Schema.  Having multiple ways to express the same constraint is useful because different ways can make sense in different contexts.  It's like trying to define the "right" way to increment a number variable in many programming languages.  In C, you can do any of these:

- `x = x + 1;`
- `x++;`
- `x += 1;`

It doesn't mean that any of them are more or less right than the others.  There's a reason that all of these exist, and there's also something to be said of code style.

That said, I think that while this works

```json
{
  "patternProperties": {"<regex>": true},
  "additionalProperties": false
}
```

it hides the _intent_ that I want all of the property names to match that regex.  Conversely

```json
{
  "propertyNames": { "pattern": "<regex>" }
}
```

does express exactly that intent.

I don't know that removing `propertyNames` is the right course of action.

**Reply by SorinGFS** [reply]
> While low usage could be an argument against introducing a proposal into the spec, it absolutely does not warrant removal from the spec.

I absolutely agree with that. And I also didn't present that low usage as an argument for removal, but as extra information about the consequences of removal.

My problem with this keyword is that it does not fit into the algorithm and for this reason it is necessary to introduce a new layer of checks on all other words.  At the same time it also breaks the annotation system, and for that another layer of checks must be introduced. Each new level of complexity reduces the performance, and in the case of the existence of this keyword the gain is so small...

The existence of this keyword also has a massive problem of a logical nature: what is the meaning of its use while a schema for what is expected in those properties cannot be added?

I have completed the validator for my implementation in which the annotation system is native. There the annotation system is used for scaling, everything is thought to be introduced based on previously added schemas. I don't want to risk publishing before adopting all the relevant tests, but I am convinced that when it can be seen it will radically change the perspective from which keywords are seen.

**Reply by jdesrosiers** [reply]
> > > Many keywords require checking instance data type before they are used - what's different?
> >
> > I finished my implementation without any data type check, except for this keyword. Not published yet. This is how I actually find the problem.
>
> This is confusing to me. Many keywords state that they only apply to a particular data type, and many tests of the JSON Schema Test Suite check that these keywords have no effect on other data types.

@SorinGFS I think we need more explanation from you on this point in order to understand your claim that `propertyNames` doesn't fit the general algorithm. The vast majority of all keywords, apply only to a given type. `properties` applies to objects, `items` applies to arrays, and `maxLength` applies to strings to name a few examples. Unless I misunderstand, it sounds like you're saying that `propertyNames` is odd because it only applies to objects. I feel like that's probably not really what you're trying to say because it's so clearly false. It doesn't seem possible to implement a keyword that applies only to a specific type without doing a data type check.

You also suggest that because `propertyNames` requires a data type check, it forces other applicator keywords to also have a data type check. I don't see how that could be. Some keywords have dependencies like `additionalProperties` being dependent on `properties`, but `propertyNames` has no dependencies and should have no effect on the behavior or implementation of other keywords.

**Reply by SorinGFS** [reply]
> it sounds like you're saying that `propertyNames` is odd because it only applies to objects. I feel like that's probably not really what you're trying to say because it's so clearly false. It doesn't seem possible to implement a keyword that applies only to a specific type without doing a data type check.

If we have a schema containing let's say... `properties`, that only applies to an object and we can assume that the data is an object. If data is not an object.. and the implementation tries to apply the schema to its inexistent keys ... the result would be an error. That error would rather indicate a schema missconfiguration, because if along with `properties` a `type: object` would exist (in my implementation `type` among all the other in-place checks are placed before the others) then the implementation would have no reason to perform another type check.

When it commes to `propertyNames` we have to apply the corresponding schema to object keys instead of values, and here is the problem: the dataLocation to be tested shares the same address as the value, the only way to apply the schema is to recurse over the keys by providing the keys themselvs as data instead of the address, and then the parser would "detect" the intent by checking the data type, which otherwise would not be neccessary.

**Reply by SorinGFS** [reply]
...continuation

Now, going through the `test-suite` tests at `properties` I saw this:

schema:
```json
            "properties": {
                "foo": {"type": "integer"},
                "bar": {"type": "string"}
            }
```
test:
```json
            {
                "description": "ignores other non-objects",
                "data": 12,
                "valid": true
            }
```

Please re-think if this is normal, a missconfigured schema which would validate wrong data! Do you think would this ever be a legitimate expectation of a schema author? I would add to this an example of yours where you are concerned about `misspelled` properties... and mean while you arn't concerned about validating wrong schema and data?

**Reply by jdesrosiers** [reply]
> If data is not an object.. and the implementation tries to apply the schema to its inexistent keys ... the result would be an error.

> Now, going through the `test-suite` tests at `properties` I saw this:

As you saw in the test you found, that's not the intended behavior. Whether that behavior is a good idea or a bad idea, it's a behavior that JSON Schema has had since the beginning (long before any of us were involved). We can't really change something with that kind of history and stability, so it wouldn't be productive to argue whether it's a good idea or not. I suggest we avoid that unnecessary argument.

**Reply by SorinGFS** [reply]
> I suggest we avoid that unnecessary argument.

Agree. Got in there to answer a question, and to highlight that the only keyword that is really coercing the artificial type check is `propertyNames`.

**Reply by jdesrosiers** [reply]
I understand some of your explanation of why `propertyNames` is checking the data type, but some parts aren't making sense to me.

> the only way to apply the schema is to recurse over the keys by providing the keys themselvs as data instead of the address

I assume you mean "iterate" rather than "recurse". This just needs be a simple loop over the object's keys. No recursion necessary.

It sounds like you made a design decision to pass JSON Pointers to your keyword handing functions rather than the actual value to be validated (or some representation of the actual value to be validated). I can see how that would be a problem, but I would consider that a problem with the design decision rather than `propertyNames`.

> , and then the parser would "detect" the intent by checking the data type, which otherwise would not be neccessary.

This part I'm not understanding. Allow me to walk through the problem as I understand it. I assume you have a function that validates a schema something like `validateSchema(schemaPointer, instancePointer)`. When calling that function in your `propertyNames` handler, you would have something like `validateSchema("/propertyNames", "/foo")`. The problem you have is that `validateSchema` doesn't know that `/foo` is pointing to the key rather than the value. It sounds like you're saying you check the data type of `/foo` to determine if you should validate against the key or the value, but I don't see how the type of `/foo` helps you decided which value to validate. You could get the keyword from the schemaPointer and decide to use the key if the keyword is `propertyNames`. Or you could pass a flag to `validateSchema` that indicates that you intend the key to be validated. Those are both just as much ugly hacks as a type check, but at least I can see how they could be a solution. I don't see how the type check solves the problem.

**Reply by SorinGFS** [reply]
> It sounds like you made a design decision to pass JSON Pointers to your keyword handing functions rather than the actual value to be validated

your assumption is correct.

> I assume you mean "iterate" rather than "recurse". This just needs be a simple loop over the object's keys. No recursion necessary.

Due to the previous answer it is a recursion. First the function finds that in schema there is `schema.propertyNames` then instead of continuing as usual by providing the address, to the function should get the object keys and pass them as string data. Then the type check is betweeen string (key names) and object (the regular argument to the function which would have the data retrieved by jsonPointer).

> It sounds like you're saying you check the data type of /foo to determine if you should validate against the key or the value, but I don't see how the type of /foo helps you decided which value to validate.

The regular argument is `jsonPointer` starting with `#`. Is a simple check if string starts with `#` or not.

> You could get the keyword from the schemaPointer and decide to use the key if the keyword is propertyNames. Or you could pass a flag to validateSchema that indicates that you intend the key to be validated. Those are both just as much ugly hacks as a type check, but at least I can see how they could be a solution.

Yea, any of these would be ugly, especially because whatever would be the "solution" would have to be added to all of those "wildcard" keywords like `additionalProperties, items, so on...`. I quoted the "solution" word because is odd! It sound like I would say: I got fat, I have 100lbs extra weight now... so the "solution" is not to walk but to take the bus! While the main problem is not that, the problem is that `propertyNames` shouldn't even exist so we wouldn't need a "solution"!

> but I would consider that a problem with the design decision rather than propertyNames.

so no! The problematic design decision is... in `json-schema`. The `propertyNames` keyword is not part of the algorithm. A problematic design decision would be to add exceptions to the rules, e.g., when adding a new keyword in `json-schema` that doesn't fit in a set of rules, or... a problematic design decision would be... not having a set of rules when adding a new keyword! This problem goes deep, very deep, to the question 'what actually is `json-schema`'? Does json-schema have an algorithmic or a lexical approach in trying to provide a common language for data exchange using `json`?

I don't know what your decision would be regarding this keyword, but do you see how many discussions, how many interpretations a single keyword that shouldn't even exist can generate? If there were clear principles that must be respected, the decision would have been simple, clear, and instant: the keyword does not fit in the algorithm, out! Period!

**Reply by jdesrosiers** [reply]
I'm trying my best to understand, but I'm still not following why recursion is necessary, what the type check is for, or how this affects other applicators.

You're right that there's no official algorithm or set of architectural principles for JSON Schema and ideally there should be. There have been attempts to define this kind of thing, but this can be a controversial subject and there's never been agreement on anything that's been proposed. As you can see from this discussion, there are a lot of different opinions on this subject and it's hard to get people to agree.

I don't understand the algorithm you're using. I do know that I didn't have the problems implementing `propertyNames` that you are having. It fits perfectly fine into my architecture without any workarounds or hacks. I haven't heard of anyone else having the problems you are either. So, I really can't agree that `propertyNames` doesn't fit the algorithm when no one else is having the same problems.

I don't see this discussion going anywhere productive at this point because there isn't consensus on what "the algorithm" is or should be. If you want to continue this discussion, I suggest starting a new discussion to propose an official algorithm for JSON Schema evaluation and try to get consensus on that topic. I don't think that's going to be easy, but if we can get that settled, we can easily resolve whether `propertyNames` belongs or not.

**Reply by SorinGFS** [reply]
>  I do know that I didn't have the problems implementing propertyNames that you are having. It fits perfectly fine into my architecture without any workarounds or hacks.

First, basically I do not have a problem in understanding `propertyNames` or including it into my implementation. In the early days I had it inside, as I described in this document. Then, I realized that by introducing that artificial type check on `additionalProperties, items, so on...` basically I validate something that isn't in the schema! So, if the schema doesn't say `type: object` why would I perform those checks on `additionalProperties, items, so on...` only if data is an object? This way, I would allow the schema author not to write that `type: object` into the schema and still do the expected validation for him.
By doing my implementation I understood few things:
- the implementation should not perform things that are not written in the schema
- there is an algorithm, and many of its parts are already represented by keywords. I'm trying to identify all its parts, I think is finite, simple, and when all its parts will have the corresponding keyword basically that would be what I think `json-schema` should be, as language or standard. Any extension to that would be possible to be made by using those basic parts of the algorithm. I will not enter off topic here, but I will open a new discussion when I'll be prepared for that. For now that is a work in progress.
- the annotation locations are uri's visited by the parser and just by having dataLocation and schemaLocation we should be able to know what operation happened there. The keyword `propertyNames` is the only keyword that cannot produce unique dataLocation uri's. This is because the keyword definition for `propertyNames` is a `wildcard`, can be anything, so when the parser is entering this keyword it can't produce a unique dataLocation uri, but instead it would overwrite the result of the previous rule. Plus, the overall result of schema validation for that dataLocation would overwrite whatever was written for `propertyNames`. This lead us to annotate multiple records on behalf of a single dataLocation, but the problem with this is that... it can't be written in a tree structure like my implementation where annotation system is native. Suddenly I had an idea how to make you understand where the problem is, and I used your implementation for that:

schema:
```json
{
  "$id": "https://json-schema.hyperjump.io/schema",
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "propertyNames": { "pattern": "^(foo|bar|baz|moo|boo|bim|doo)$" }
}
```

data:
```json
{"foo":{},"fooo":{},"bar":{},"baz":{}}
```

the result:
- Invalid
    - fails schema constraint https://json-schema.hyperjump.io/schema#/propertyNames

The question is: which one? 😄 
Now, if you will try to fix this and give the right annotation in order to be used for debug or programmatic use you will be able to see what I meant about unique dataLocation in annotations.

**Reply by gregsdennis** [reply]
I think I see what you're trying to say here.

Let's look at this schema

```json
{
  "type": "array",
  "items": {
    "type": "object",
    "propertyNames": { "enum": [ "foo", "bar", "baz" ] }
  }
}
```

and this instance

```json
[
  { "foo": 1 },
  { "disallowed": 2 },
  { "bar": 3 },
  { "baz": 4 },
]
```

My implementation (what's currently deployed to https://json-everything.net/json-schema) produces this output unit (among others):

```json
{
  "valid": false,
  "evaluationPath": "/items",
  "schemaLocation": "https://json-everything.net/3cce902220#/items",
  "instanceLocation": "/1",
  "details": [
    {
      "valid": false,
      "evaluationPath": "/items/propertyNames",
      "schemaLocation": "https://json-everything.net/3cce902220#/items/propertyNames",
      "instanceLocation": "/1/disallowed",
      "errors": {
        "enum": "Value should match one of the values specified by the enum"
      }
    }
  ]
}
```

I think what you're saying is that the first output unit inadequately describes the problem that `propertyNames` failed _and_ which properties it failed on.

I counter this argument by inferring that `propertyNames` failed on the "disallowed" property because of the information in the detail output unit.  Granted, the `instanceLocation` (not `dataLocation`) isn't technically correct here because the subschema at `/items/propertyNames` isn't validating the instance _value_ at `/1/disallowed` but rather the property name "disallowed".  However, I chose this approach because the inference of having `propertyNames` in `schemaLocation`/`evaluationPath` gives us the context that we're looking at property names, not values.

That said, an error message in the outer output unit could make things a little clearer:

```json
{
  "valid": false,
  "evaluationPath": "/items",
  "schemaLocation": "https://json-everything.net/3cce902220#/items",
  "instanceLocation": "/1",
  "errors": {
    "propertyNames": "Properties [\"disallowed\"] failed validation"
  },
  "details": [
    {
      "valid": false,
      "evaluationPath": "/items/propertyNames",
      "schemaLocation": "https://json-everything.net/3cce902220#/items/propertyNames",
      "instanceLocation": "/1/disallowed",
      "errors": {
        "enum": "Value should match one of the values specified by the enum"
      }
    }
  ]
}
```

But error messaging is not specified by the output (on purpose).  (_How to include_ an error message is specified, but not the messages themselves or even what keywords should generate them.)

I'm interested to know if you have an alternate solution to this problem.

---

As far as not fitting into "the algorithm", I'm not sure that that can be quantified.  JSON Schema doesn't define an algorithm.  And it avoids doing so on purpose.  It defines behaviors: output for a given input.  How an implementation gets there is up to the implementer.

It's intentionally permissive to allow you to determine what the best design choices are for your consumers.

I can't see that JSON Schema will ever specify an algorithm because the project values this freedom for the implementers.

**Reply by SorinGFS** [reply]
> I'm interested to know if you have an alternate solution to this problem.

Therein lies the problem: there is no straight forward solution. And this keyword is the only one with this problem. An implementation, depending on proccessing order may even validate an `object[key]` before failing to validate the `key`, therefore leading to a greater confusion when would produce different validation results for the same `instanceLocation` (as you call it). Any other keyword by `instanceLocation` will mean the value of that key, while `propertyNames` means the key (by recursing over object keys one by one). That's why I said that `propertyNames` doesn't fit in the same algorithm as all the other keywords. A schema can be a boolean (not the case) or an object. A proper algorithm for `propertyNames` would have the `json` objects represented as nested arrays with two elements like this:

```json
[ "foo", [ "bar", [ "baz", 3 ] ] ]
```

where accessing `arr[1][1][0]` we retrieve the object's `propertyName` and by accessing `arr[1][1][1]` we retrieve the object's `propertyValue`. That algorithm would provide unique paths for all the keys and all the values. The `json-schema` algorithm provides unique paths only for the values.

To overcome this problem every implementation in it's own programming language must solve this in its own way, but while in `C#` you will not feel the performance impact since a `.dll` or `.so` works I don't know, 20, 50 times faster the others will feel any unneccessary added thing. The performance impact for `dynamicRef`, or for `unevaluated*` keywords are IMO justified, but not for `propertyNames`.

An important detail is the following: the `instanceLocation` produced in annotation is (IMO) not important to be understandable for a human eye, but for a programming language in order to use it for example to indicate in an interface where the problem is, or similar things.

> As far as not fitting into "the algorithm", I'm not sure that that can be quantified. JSON Schema doesn't define an algorithm. And it avoids doing so on purpose. It defines behaviors: output for a given input. How an implementation gets there is up to the implementer.
> It's intentionally permissive to allow you to determine what the best design choices are for your consumers.
> I can't see that JSON Schema will ever specify an algorithm because the project values this freedom for the implementers.

I think this is a discussion that should exist, not to limit the freedom for the implementers but quite the oposite of that. I think if `json-schema` would be seen as an algorithm in which keywords are the lexical representation of a concern in the algorithm and in which every part of the algorithm is represented by a stable keyword, then that would be where `json-schema` ends and implementation is starting. For example, let's consider the keyword `allOf`, as part of the algorithm is an array with at least one positional element (`prefixItems`). As part of an implementation vocabulary can mean different things: implementation A may say `validate` all the items, implementation B may say `iterate` all the items and print them in the console, each implementation would have their own meta schemas based on `json-schema` algorithm, where they can decide which keywords are using in that specific vocabulary, and it will be their own responsability to document the specifications of their project. When users would choose a meta schema url they will search for the corresponding documentation there. I just wanted to give a basic idea here, I don't want to bring this off topic, as I said to Jason I will bring the full vision of this when time arives.

**Reply by jdesrosiers** [reply]
> Now, if you will try to fix this and give the right annotation in order to be used for debug or programmatic use you will be able to see what I meant about unique dataLocation in annotations.

I understand that part. I agree that's a problem. I acknowledged it from the beginning of the discussion. The part I don't understand is the consequences you claim that has on implementing JSON Schema. The only consequence I see is with the output format which can't precisely report errors/annotations for this keyword because it relies on JSON Pointer to identify the location of the error/annotation. I've acknowledged that maybe that's an indication that `propertyNames` doesn't belong. What I don't understand is how that can have the impact on implementation that you're describing. I personally see no impact at all on implementation. The only impact is in reporting. So, I'd rather explore if we can solve the reporting issue before considering removing the keyword.

> An important detail is the following: the `instanceLocation` produced in annotation is (IMO) not important to be understandable for a human eye, but for a programming language in order to use it for example to indicate in an interface where the problem is, or similar things.

I want to also acknowledge that I agree with this point. But, again, I'd like to first try to solve this as an issue with the output format before considering the much more extreme solution of removing the keyword.

**Reply by SorinGFS** [reply]
Ok.

**Comment by SorinGFS** [comment]
@all

what I saw in schemastore schemas people are using much more often the following instead of `propertyNames`:

```js
properties: {
    foo: {},
    bar: {},
    baz: false
}
```

The intent is even more clearer like that, and for the pattern cases there is `patternProperties`

**Reply by gregsdennis** [reply]
Given my experience of seeing and answering questions (Slack, GitHub, StackOverflow) over multiple years, I take this as people not knowing about `propertyNames` rather than an active refusal to use it.  When telling people about `propertyNames`, their response is generally, "Thanks! That's exactly what I needed."

I don't think that a lack of use in the wild can be used as evidence that people are _avoiding_ it.

**Reply by SorinGFS** [reply]
> When telling people about propertyNames, their response is generally, "Thanks! That's exactly what I needed."

And that is my existential question regarding `json-schema`: is `json-schema` a service that satisfies people's needs, or is it an entity that tries to standardize data transfer by having an algorithmic approach to `json`?

**Reply by gregsdennis** [reply]
(Your question has no bearing on the text you quoted.)

It's neither.  It's a standardized declarative syntax that allows people to define constraints for JSON data.

If your argument is that `propertyNames` doesn't "fit the algorithm", then I think you have a fundamental misunderstanding of how JSON Schema works, which I and others have tried to rectify (unsuccessfully, it seems).

**Reply by SorinGFS** [reply]
> When telling people about propertyNames, their response is generally, "Thanks! That's exactly what I needed."

When you give me that as an argument for the existence of the keyword, I am speechless, rest my case, no further questions!

But you don't limit yourself to that: you also unilaterally decide to close the discussion (it's not the first time), and I wouldn't be surprised if you hide or even delete my comments! Again, it wouldn't be the first time. 
It's convenient, isn't it, when something doesn't suit us we delete it and behave as if it didn't exist :))

I noticed a massive "purification" of the discussions from the past, as if the discussions of the last months never took place!

If you think this dicatorial decision mechanism is beneficial for the future `json-schema`... I don't know what else I could say...

**Reply by gregsdennis** [reply]
> I wouldn't be surprised if you hide or even delete my comments

I haven't hidden anything.  In the past, you've taken discussions off-topic, but you've remained on topic here.

> you also unilaterally decide to close the discussion

I (as a maintainer of the project) see no reason to remove a feature that (1) is doing no harm and (2) is actually useful.  As such, the proposal to remove it is being rejected.  The discussion is concluded, and _someone_ needs to close it, be it Jason, myself, or any of the other maintainers of the project.

As I mentioned, I didn't lock the conversation, so you're welcome to continue providing evidence that your proposal should be reconsidered, but for now, the matter is decided.

> I noticed a massive "purification" of the discussions from the past, as if the discussions of the last months never took place!

I (as a maintainer of the project) have recently been performing some house cleaning.  I have _not_ deleted or hidden any issues, discussions, or comments on them.  They're all still available.  What I have done is update how we're organizing issues _in the spec repo_ so that we can better track multiple lines of work.

Over a month ago I mentioned [in Slack](https://json-schema.slack.com/archives/CT7FF623C/p1718701438506199) that I was planning on doing this, and then I created [an issue](https://github.com/json-schema-org/json-schema-spec/issues/1526) in the spec repo to record my findings and make an official proposal.  Reception to my proposal was good, so I carried it out today.

I can understand that such a change can be disruptive, but (to be frank) it's a listing of the work _we_ need to do, and _we_ are the ones that have to manage it.

**Reply by gregsdennis** [reply]
> you also unilaterally decide to close the discussion

Yes.  @SorinGFS you are correct here, and I apologize.  I should have called for a vote to close it.

I feel it's a bit awkward to reopen this only to call for a vote to close it.  Would you be satisfied with a promise not to act unilaterally in the future?

**Reply by SorinGFS** [reply]
> Would you be satisfied with a promise not to act unilaterally in the future?

Yes, sure.

**Comment by gregsdennis** [comment]
To me, a conversation about removing any keyword needs to consider both its longevity and how complete its definition is.

`propertyNames` is a long-standing and well-understood keyword, even if newer users [just don't know it's there](https://github.com/orgs/json-schema-org/discussions/768#discussioncomment-10080015).  Its inclusion in the specification has not caused any problems or confusion for users or implementers. (@SorinGFS you're the first that I've seen complain about this keyword.)

Given that, I don't think the arguments presented warrant removal from the specification.  I'm going to close this discussion, but it will remain unlocked.
